### PR TITLE
`omitempty` corrections

### PR DESCRIPTION
**Pull request by Tom-Newton**
## Purpose of this PR
Some corrections to make a couple of required fields required and optional fields genuinely optional.

**Proposed changes:**
- Fix https://github.com/kubeflow/spark-operator/issues/2215. This problem was caused by some optional fields in the CRD not using `omitempty`. That meant that valid `SparkApplication` configs got `null`s added were they weren't allowed during some encode and decode in the SparkApplication default webhook. 
- Use regex `\+optional\n^(?!.*omitempty).*$` to find all the optional parameters that don't use `omitempty` and add `omitempty` to them, to avoid the same problem in other places. 
- Remove `omitempty` on `metadata` and `spec` fields of `SparkApplication` and `ScheduledSparkApplication`. I think these should be required and they used to be in the [past](https://github.com/kubeflow/spark-operator/pull/2072/files#diff-1ac2379092396c2724760cdb762efd6dfef8bb5122a29df60cd1da949b306a54L157-L158). and I suspect this may have been changed accidentally. Obviously let me know if there is good reason to add  `omitempty` on these. 
- Make `mainApplicationFile` a required field
- ~~Write unittests for both of the above. I wasn't really sure how to do this but I came up with something. I don't know if its a good way to do it, but it adds the required coverage. My approach was to read the openapiv3 schema from the CRD and validate against that.~~ Ended up removing this maybe we can add something better as a follow up. 

## Change Category
Indicate the type of change by marking the applicable boxes:

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [ ] Documentation update

## Checklist
Before submitting your PR, please review the following:

- [x] I have conducted a self-review of my own code.
- [x] I have updated documentation accordingly - Ran `make build-api-docs` to capture making `mainApplicationFile` required. 
- [x] I have added tests that prove my changes are effective or that my feature works.
- [x] Existing unit tests pass locally with my changes.

### Additional Notes
I'm still a `golang` noob.

### CODE REVIEW

**Comment by ChenYi015** [comment]
> Remove omitempty on metadata and spec fields of SparkApplication and ScheduledSparkApplication. I think these should be required and they used to be in the [past](https://github.com/kubeflow/spark-operator/pull/2072/files#diff-1ac2379092396c2724760cdb762efd6dfef8bb5122a29df60cd1da949b306a54L157-L158). and I suspect this may have been changed accidentally. Obviously let me know if there is good reason to add omitempty on these.

@Tom-Newton Since v2, we use [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) to create api for SparkApplication/SchedulerSparkApplication with the command like follows:

```bash
kubebuilder create api --version v1beta2 --kind SparkApplication
```

and the `omitempty` annotation exists by default.

**Comment by Tom-Newton** [comment]
> @Tom-Newton Since v2, we use [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) to create api for SparkApplication/SchedulerSparkApplication with the command like follows:
> 
> ```shell
> kubebuilder create api --version v1beta2 --kind SparkApplication
> ```
> 
> and the `omitempty` annotation exists by default.

Thanks for the info. Does that mean that mean we should keep them as they were? or is there somewhere else that is source of truth to mark them as required when running `kubebuilder`?

**Comment by ChenYi015** [comment]
> Does that mean that mean we should keep them as they were? 

I think it is better to keep the metadata and spec annotations as default unless there is a good reason.  For many other types, such as the ones in the [kubeflow/training-operator](https://github.com/kubeflow/training-operator), the `omitempty` annotation is retained:

https://github.com/kubeflow/training-operator/blob/6965c1a92462d46981071748936eb135a4584f3d/pkg/apis/kubeflow.org/v1/pytorch_types.go#L56-L68

**Comment by ChenYi015** [comment]
The serviceType in driverIngressConfigurations also need to add `omitempty` annotations:

https://github.com/kubeflow/spark-operator/blob/1491550391f224224b621fa64fc2f3e05a4cbe2a/api/v1beta2/sparkapplication_types.go#L313-L315

File: internal/webhook/sparkapplication_defaulter_test.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
+/*
```
Comment (ChenYi015) [inline internal/webhook/sparkapplication_defaulter_test.go:1]: I think the `sparkapplication_defauler_test.go` file should be used to implement unit tests against funcs related to `SparkApplicationDefaulter`, such as `defaultSparkApplication`. We do not need to instantiate the webhook manually and call the handle func. Maybe we can remove this file in this PR and later add a new e2e test in another PR to validate that the serialize/deserialize process can work as expected.

**Comment by Tom-Newton** [comment]
> The serviceType in driverIngressConfigurations also need to add `omitempty` annotations:
> 
> https://github.com/kubeflow/spark-operator/blob/1491550391f224224b621fa64fc2f3e05a4cbe2a/api/v1beta2/sparkapplication_types.go#L313-L315

Yes, I think I got that one https://github.com/kubeflow/spark-operator/pull/2255/files#diff-fae12edea9174bb072c60830180dc6e65aeaf098de9dfb7ac69240c3e1c347b1R315

**Comment by Tom-Newton** [comment]
> I think it is better to keep the metadata and spec annotations as default unless there is a good reason. 

My reasoning is:
1. It seems wrong to make something optional in the CRD schema when we know its actually required. Currently if I submit a `SparkApplication` with no `spec` it fails in the defaulting webhook with `* spec.type: Unsupported value: "": supported values: "Java", "Python", "Scala", "R"`. This is better than just admitting it and causing a `panic` in the controller but it does show that `spec` is a required field really. 
1. Other code that iteracts with `SparkApplication`s using the CRD schema now has to add extra checks for whether these fields are `null`. We have a case like this.

Let me know if you think these constitute good reason - if not I'll revert them.

**Comment by ChenYi015** [comment]
> > I think it is better to keep the metadata and spec annotations as default unless there is a good reason.
> 
> My reasoning is:
> 
> 1. It seems wrong to make something optional in the CRD schema when we know its actually required. Currently if I submit a `SparkApplication` with no `spec` it fails in the defaulting webhook with `* spec.type: Unsupported value: "": supported values: "Java", "Python", "Scala", "R"`. This is better than just admitting it and causing a `panic` in the controller but it does show that `spec` is a required field really.
> 2. Other code that iteracts with `SparkApplication`s using the CRD schema now has to add extra checks for whether these fields are `null`. We have a case like this.

@Tom-Newton

1. When creating a SparkApplication with no spec, the CR will not be written into etcd because the API server fails to validate the CRD schema, thus webhook will not be called and no events are sent to the controller.

2. Since `metadata` and `spec`  fields are structs, not pointers, so there is no need to consider the `null` problem when interacting with the CRD schema.

**Comment by Tom-Newton** [comment]
> When creating a SparkApplication with no spec, webhook will not be called because the API server will first fail to validate the CRD schema. 

From my testing that is not correct. If I use the use the following config
```
apiVersion: sparkoperator.k8s.io/v1beta2
kind: SparkApplication
metadata:
  name: test
  namespace: default
```
and I test against `kind` running the spark operator built from the latest `master`. Then I get 
```
kubectl apply -f ~/Downloads/config.yml 
The SparkApplication "test" is invalid: 
* spec.type: Unsupported value: "": supported values: "Java", "Python", "Scala", "R"
* spec.mainApplicationFile: Invalid value: "null": spec.mainApplicationFile in body must be of type string: "null"
```
and I see from the webhook logs that this did call the webhook
```
2024-10-18T09:30:08.421Z	INFO	webhook/sparkapplication_defaulter.go:57	Defaulting SparkApplication	{"name": "test", "namespace": "default", "state": ""}
``` 

> the API server will first fail to validate the CRD schema. 

This is exactly what I aim to achieve by removing the `omitempty`. With the `omitempty` in place the generated CRD schema will not mark the `spec` as required. However after removing the `omitempty` and re-running `make update-crd` (as I've done in this PR) then the CRD now does mark the `spec` as required https://github.com/kubeflow/spark-operator/pull/2255/files#diff-929d75537a6ddea49aa2609c3f42e8bf15c99f31a82aae9b8c5a7933d78199ccR11558-R11560.

**Comment by ChenYi015** [comment]
I have just tested it and you are right. The webhook is called and the `metadata` and `spec` fields become required after removing the `omitempty` annotation.

File: api/v1beta2/sparkapplication_types.go [hunk 2]
```diff
@@ -63,7 +63,7 @@ type SparkApplicationSpec struct {
 	MainClass *string `json:"mainClass,omitempty"`
 	// MainFile is the path to a bundled JAR, Python, or R file of the application.
 	// +optional
-	MainApplicationFile *string `json:"mainApplicationFile"`
+	MainApplicationFile *string `json:"mainApplicationFile,omitempty"`
```
[the remark this answers is not in the document]
Comment (Tom-Newton) [reply api/v1beta2/sparkapplication_types.go:66]: Quite possibly. I was quite surprised to see it was optional

[the remark this answers is not in the document]
Comment (ChenYi015) [reply api/v1beta2/sparkapplication_types.go:66]: Let's mark this field as required and this PR will LGTM.

[the remark this answers is not in the document]
Comment (Tom-Newton) [reply api/v1beta2/sparkapplication_types.go]: Done

**Review by ChenYi015 — APPROVED** [review]

**Comment by ChenYi015** [comment]
@Tom-Newton Well done! Thanks for reporting the issue and fixing it.
/lgtm

**Comment by Tom-Newton** [comment]
> @Tom-Newton Well done! Thanks for reporting the issue and fixing it. 

Thanks for reviewing :slightly_smiling_face:

File: internal/webhook/sparkapplication_defaulter_test.go (hunk 1 above)
Comment (Tom-Newton) [reply internal/webhook/sparkapplication_defaulter_test.go]: I created an issue so ths doesn't get forgotten https://github.com/kubeflow/spark-operator/issues/2284

**Pull request merged**